### PR TITLE
rgw: Fix bug on subuser create for user_id with colon

### PR DESCRIPTION
**Pull request by clwluvw**
user_id with colon like foo:bar can't create subuser with radosgw-admin subuser create without uid given
```
radosgw-admin subuser create --subuser foo:bar:subuser
```

Fixes: https://tracker.ceph.com/issues/45236
Signed-off-by: Seena Fallah seeenafallah@gmail.com

### CODE REVIEW

**Comment by theanalyst** [comment]
Is there an issue here that an explicit uid and subuser doesn't fixa?

ie. 
radosgw-admin subuser create --uid=foo:bar --subuser baz 

this would correctly create foo:bar:baz as a subuser

**Comment by clwluvw** [comment]
> this would correctly create foo:bar:baz as a subuser

@theanalyst Yes but in a form of `--subuser foo:bar:baz` it will fails!

**Comment by theanalyst** [comment]
> > this would correctly create foo:bar:baz as a subuser
> 
> @theanalyst Yes but in a form of `--subuser foo:bar:baz` it will fails!

Understood, the current behaviour is this would create a user bar:baz under uid foo, and you're moving this to foo:bar with subuser baz. But ideally in these cases an explicit --uid would make things clearer, which is the documented way: 

https://docs.ceph.com/docs/master/radosgw/admin/#create-a-subuser

Maybe we could just also add a note in doc/radosgw/admin.rst that it is better to explicitly specify a `--uid` while creating a subuser, otherwise the last/first colon part would be parsed as the user.

**Comment by theanalyst** [comment]
Hi Seena, 

we discussed this in the standups today, I think an approach would be just to add a note on the current behaviour in the admin docs, as in these cases being explicit is always better (ie. specifying both --uid and --subuser).

The only change happening here is --subuser user1:sub1:sub2 would actually create sub2 under a uid user1:sub1 whereas the current behaviour is that subuser sub1:sub2 gets created under user1, since we don't really restrict colons in subuser names either the best way would be being explicit when ambigious and just documenting the current behaviour. What do you think?

**Comment by clwluvw** [comment]
> we don't really restrict colons in subuser names

@theanalyst I have tried to create a subuser with colons but it fails. Am I wrong?
```bash
radosgw-admin subuser create --subuser bar:baz --uid foo
```

**Comment by theanalyst** [comment]
> > we don't really restrict colons in subuser names
> 
> @theanalyst I have tried to create a subuser with colons but it fails. Am I wrong?
> 
> ```shell
> radosgw-admin subuser create --subuser bar:baz --uid foo
> ```

Do you have a user called foo in this case? The command creates a subuser called bar:baz under user foo, is that failing?

**Comment by clwluvw** [comment]
> > > we don't really restrict colons in subuser names
> > 
> > 
> > @theanalyst I have tried to create a subuser with colons but it fails. Am I wrong?
> > ```shell
> > radosgw-admin subuser create --subuser bar:baz --uid foo
> > ```
> 
> Do you have a user called foo in this case? The command creates a subuser called bar:baz under user foo, is that failing?

@theanalyst Yes it fails because `set_subuser` function set `user_id` bar and so there is no bar user even if we set `--uid`.

**Comment by theanalyst** [comment]
> > > > we don't really restrict colons in subuser names
> > > 
> > > 
> > > @theanalyst I have tried to create a subuser with colons but it fails. Am I wrong?
> > > ```shell
> > > radosgw-admin subuser create --subuser bar:baz --uid foo
> > > ```
> > 
> > 
> > Do you have a user called foo in this case? The command creates a subuser called bar:baz under user foo, is that failing?
> 
> @theanalyst Yes it fails because `set_subuser` function set `user_id` bar and so there is no bar user even if we set `--uid`.

Ah interesting, I didn't notice as I actually had another user with the name, this definitely needs a documentation note alongwith the changeset, can you add another commit adding a note in admin.rst about the behaviour when a subuser create is allowed without an explicit --uid that we parse the user until the last colon.

File: doc/radosgw/admin.rst [hunk 1]
```diff
@@ -100,6 +100,13 @@ For example::
   radosgw-admin subuser create --uid=johndoe --subuser=johndoe:swift --access=full
 
 
+Also you can create subuser without specifying ``--uid={username}``. ::
```
Comment (theanalyst) [inline doc/radosgw/admin.rst:103]: Can you rephrase it so, 
It is preferred to be explicit when creating subusers, specifying users with a `--uid`. If a `--uid` is not supplied, we parse the `uid` until the last colon. 

<syntax>

Comment (clwluvw) [reply doc/radosgw/admin.rst]: Sure.

**Comment by cbodley** [comment]
ambiguity in string parsing is why we shouldn't ever rely on parsing `user:subuser`

one of the suggestions from the standup was to require the `--uid` option for any subuser commands, so that we don't have to do any parsing of the `--subuser` string

**Comment by clwluvw** [comment]
> ambiguity in string parsing is why we shouldn't ever rely on parsing `user:subuser`
> 
> one of the suggestions from the standup was to require the `--uid` option for any subuser commands, so that we don't have to do any parsing of the `--subuser` string

I think this would be allow users to create subuser with colon! Is it okay to add this ability? I don't think now there is a ability to create a subuser with colon.

**Comment by clwluvw** [comment]
Ping @theanalyst

**Comment by clwluvw** [comment]
unstale please

**Comment by clwluvw** [comment]
unstale please

File: doc/radosgw/admin.rst [hunk 2]
```diff
@@ -100,6 +100,12 @@ For example::
   radosgw-admin subuser create --uid=johndoe --subuser=johndoe:swift --access=full
 
 
+It is preferred to be explicit when creating subusers, specifying users with a ``--uid``.
+If a ``--uid`` is not supplied, we parse the uid until the last colon ::
+
```
Comment (anthonyeleven) [inline doc/radosgw/admin.rst:105]: I think a better phrasing might be

"When creating a subuser, it is best to explicitly indicate the parent user with a ``--uid`` argument.  If ``--uid`` is not supplied, the command expects the ``--subuser`` value to be prefixed with the parent user followed by a colon:

**Review by anthonyeleven — CHANGES_REQUESTED** [review]
Please update the commit and PR messages to something like

"doc/radosgw: Clarify subuser creation with radosgw-admin"

Apologies for this PR falling through the cracks.  We have updated processes to help avoid this in the future and are going back to handle older PRs.

File: doc/radosgw/admin.rst (hunk 2 above)
Comment (clwluvw) [reply doc/radosgw/admin.rst:105]: @anthonyeleven It's not only about doc! As you can see the diff it's changing the current cli behavior.

**Comment by mattbenjamin** [comment]
@cbodley what shall we do with this one?
